atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/CoordinationService.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.coordination;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+import com.palantir.atlasdb.keyvalue.impl.CheckAndSetResult;
+
+/**
+ * A {@link CoordinationService} is used to agree on values being relevant or correct at a given timestamp.
+ * The sequence of values being agreed should evolve in a backwards consistent manner; that is, if I read a value
+ * for some timestamp TS and make a decision at TS, all future values written to the {@link CoordinationService} should
+ * also ensure that no decision that would be inconsistent with the decision made at TS would be made at TS. In general,
+ * this means that one should avoid changes that affect behaviour of values below the validity bound.
+ *
+ * It is the responsibility of the caller to preserve the above property.
```

Comment:
How is this done? I'd assume that the service/store layer enforce that you can't change behavior for some old bound, so it's up to clients of this service to respect the behavior communicated for each bound.